controller/internal/enforcer/envoyproxy/authz/server.go

```diff
@@ -0,0 +1,501 @@
+		zap.L().Debug("ext_authz_server: service info", zap.String("service", serviceName), zap.Any("info", info))
+	}
+
+	// TODO: figure out why an abstract unix socket path doesn't work
+	// Create a custom listener
+	//addr, err := net.ResolveUnixAddr("unix", s.socketPath)
+	//if err != nil {
+	//	return nil, err
+	//}
+	//nl, err := net.ListenUnix("unix", addr)
+	addr, err := net.ResolveTCPAddr("tcp", s.socketPath)
+	if err != nil {
+		return nil, err
+	}
+	nl, err := net.ListenTCP("tcp", addr)
+	if err != nil {
+		return nil, err
+	}
+
+	// start and listen to the server
+	go s.run(nl)
+
+	return s, nil
+}
+
+func (s *Server) run(lis net.Listener) {
+	zap.L().Debug("Starting to serve gRPC for ext_authz server", zap.String("puID", s.puID), zap.String("direction", s.direction.String()))
+	if err := s.server.Serve(lis); err != nil {
+		zap.L().Error("gRPC server for ext_authz failed", zap.String("puID", s.puID), zap.Error(err), zap.String("direction", s.direction.String()))
+	}
+	zap.L().Debug("stopped serving gRPC for ext_authz server", zap.String("puID", s.puID), zap.String("direction", s.direction.String()))
+}
+
+// Stop calls the function with the same name on the backing gRPC server
+func (s *Server) Stop() {
+	s.server.Stop()
+}
+
+// GracefulStop calls the function with the same name on the backing gRPC server
+func (s *Server) GracefulStop() {
+	s.server.GracefulStop()
+}
+
+// Check implements the AuthorizationServer interface
+func (s *Server) Check(ctx context.Context, checkRequest *ext_authz_v2.CheckRequest) (*ext_authz_v2.CheckResponse, error) {
+	switch s.direction {
+	case IngressDirection:
+		return s.ingressCheck(ctx, checkRequest)
+	case EgressDirection:
+		return s.egressCheck(ctx, checkRequest)
+	default:
+		return nil, fmt.Errorf("direction: %s", s.direction)
+	}
+}
+
+// ingressCheck implements the AuthorizationServer for ingress connections
+func (s *Server) ingressCheck(ctx context.Context, checkRequest *ext_authz_v2.CheckRequest) (*ext_authz_v2.CheckResponse, error) {
+	// TODO: needs to be removed before we merge: this exposes secret data, and must never be in real logs - even in the debug case
+	zap.L().Debug("ext_authz ingress: checkRequest", zap.String("puID", s.puID), zap.String("checkRequest", checkRequest.String()))
+
+	// get the PU context
+	pctxRaw, err := s.puContexts.Get(s.puID)
+	if err != nil {
+		zap.L().Error("ext_authz ingress: failed to get PU context", zap.String("puID", s.puID), zap.Error(err))
+		return createDeniedCheckResponse(rpc.INTERNAL, envoy_type.StatusCode_InternalServerError, "failed to get PU context"), nil
+	}
+	pctx, ok := pctxRaw.(*pucontext.PUContext)
+	if !ok {
+		zap.L().Error("ext_authz ingress: PU context has the wrong type", zap.String("puID", s.puID), zap.String("puContextType", fmt.Sprintf("%T", pctxRaw)))
+		return createDeniedCheckResponse(rpc.INTERNAL, envoy_type.StatusCode_InternalServerError, "PU context has the wrong type"), nil
+	}
+
+	// extract our headers
```

Comment:
here you need the sctx of the app proxy. There can be multiple services listening for a given PU. I am not sure if we can recover the original destination/port from the request or how envoy will provide this.

Reply:
We obviously need to add the user claims from the Authorization Bearer header. 

I am noticing lots of commonality with the application proxy. I think we should refactor the app proxy, and use it as a library in both places and then we don't repeat the logic twice. The processAppRequest, processNetRequest functions are pretty similar with minor modifications on what they return. If we refactor the logic out, we can use a common code based whether we are inline or not.

Reply:
I started noticing that as well, and it definitely would have been one of my proposals: we need to make parts of the app proxy reusable in places like this.

re the sctx: I'm not sure how there could be any service listening at all for a PU here: this currently does not support at all the running application proxy, nor is it started for this enforcer.

The biggest trouble here for us is that we actually would need to distinguish between Istio pods, and just plain envoy pods where we actually might want to setup iptables ourselves, and run an app proxy. In Istio this is completely impossible because they take over all the traffic.

For this enforcer in general I think we should go about it more the "developer" way: make things configurable and make the user partially responsible for making it work. This provides flexibility which is exactly what people would like to have when they choose this enforcer. If they don't want that, well just go with our standard enforcer.